Public/Src/Tools/Execution.Analyzer/Analyzers.Partner/ContentPlacementAnalyzer.cs

```diff
@@ -0,0 +1,532 @@
+            base.Dispose();
+        }
+
+        public override void FileArtifactContentDecided(FileArtifactContentDecidedEventData data)
+        {
+            string extension = Path.GetExtension(data.FileArtifact.Path.ToString(CachedGraph.Context.PathTable));
+            AddToDictionaryWithListAsValue(m_artifactsPerExtension, extension, data.FileArtifact);
+            CountInDictionary(m_countFilesPerExtension, extension);
+            m_fileContentMap[data.FileArtifact] = data.FileContentInfo;
+        }
+
+        /// <inheritdoc />
+        public override void PipExecutionPerformance(PipExecutionPerformanceEventData data)
+        {
+            AddToDictionaryWithListAsValue(m_scheduledPipExecutionData, data.PipId.Value, data);
+        }
+
+
+        private IList<ValType> GetOrReturnNull<KeyType, ValType>(Dictionary<KeyType, IList<ValType>> dictionary, KeyType key)
+        {
+            if (dictionary.ContainsKey(key))
+            {
+                return dictionary[key];
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        private void AddToDictionaryWithListAsValue<KeyType, ValType>(Dictionary<KeyType, IList<ValType>> dictionary, KeyType key, ValType value)
+        {
+            if (dictionary.ContainsKey(key))
+            {
+                dictionary[key].Add(value);
+            }
+            else
+            {
+                dictionary.Add(key, new List<ValType>(new ValType[] { value }));
+            }
+        }
+
+        private void CountInDictionary<KeyType>(Dictionary<KeyType, int> dictionary, KeyType key)
+        {
+            if (dictionary.ContainsKey(key))
+            {
+                dictionary[key] += 1;
+            }
+            else
+            {
+                dictionary.Add(key, 1);
+            }
+        }
+
+    }
+
+    public sealed class BXLPipData
+    {
+        public string SemiStableHash { get; set; } = null;
+        public int Priority { get; set; } = 0;
+        public int Weight { get; set; } = 0;
+        public int TagCount { get; set; } = 0;
+        public int SemaphoreCount { get; set; } = 0;
+        public double DurationMs { get; set; } = -1;
+        public string Type { get; set; } = null;
+        public string ExecutionLevel { get; set; } = null;
+        public int DependencyCount { get; set; } = 0;
+        public int InputCount { get; set; } = 0;
+        public int OutputCount { get; set; } = 0;
+        public long StartTimeTicks { get; set; } = 0;
+
+        public static List<BXLPipData> BuildFromExecutionData(Pip pip, IList<PipExecutionPerformanceEventData> executionData,
+            Dictionary<uint, int> scheduledPipDependencyCount,
+            Dictionary<uint, int> scheduledPipInputCount,
+            Dictionary<uint, int> scheduledPipOutputCount,
+            PathTable pathTable,
+            StringTable stringTable)
+        {
+            var output = new List<BXLPipData>();

```

Comment:
nit: small perf improvement. If you know the size of the list before hand like here, you can pass the size to the list.
That way the underlying array of List<T> is allocated to that size immediately.
i.e. `new List<BxlPipData>(executionData.Count)`

else it creates an array with 8 and when you add the 9th the list will create a new larger array, say 16, copy all entries over and drop the array with length 8. Now when you hit 17 the same will happen again over and over. Since in BuildXL we are typically working with quite large arrays this can save a lot of unneeded computations.